### PR TITLE
LinkedList implementation of TimeRangeGroup

### PR DESCRIPTION
**Pull request by talvarez0507**
This is the linked list implementation of TimeRangeGroup. It uses a linear search for hasTimeRange and edits the list in place for adding and deleting. The tests are exactly the same as the tests for the ArrayList implementation.

### CODE REVIEW

File: portfolio/src/test/java/com/google/sps/LinkedListTimeRangeGroupTest.java [hunk 1]
```diff
@@ -0,0 +1,557 @@
+package com.google.sps.data;
```
Comment (talvarez0507) [inline portfolio/src/test/java/com/google/sps/LinkedListTimeRangeGroupTest.java]: These are exactly the same tests as ArrayListTimeRangeGroupTest.java which is already in master except this tests LinkedListTimeRangeGroup instead.

**Review by hollyyuqizheng — APPROVED** [review]

**Review by raulcruise — APPROVED** [review]

File: portfolio/src/main/java/com/google/sps/data/LinkedListTimeRangeGroup.java [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+import java.time.Instant;
+import java.util.Collection;
+import java.util.Iterator;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.ListIterator;
+
+/**
+ * Models an implementation of the TimeRangeGroup model using LinkedList. This is a slightly
+ * modified version of what hollyyuqizheng wrote for the ArrayList implementation.
```
Comment (jamesnam-google) [inline portfolio/src/main/java/com/google/sps/data/LinkedListTimeRangeGroup.java]: I don't think you need to mention hollyyuqizheng. You can write that it modified ArrayListTime...

File: portfolio/src/main/java/com/google/sps/data/LinkedListTimeRangeGroup.java [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+ */
+public class LinkedListTimeRangeGroup implements TimeRangeGroup {
+
+  // This list of all the time ranges will be sorted by start time ascending
+  // in add and delete methods. The time ranges stored in this list are
+  // pair-wise disjoint at any moment.
+  public List<TimeRange> allTimeRanges;
```
Comment (jamesnam-google) [inline portfolio/src/main/java/com/google/sps/data/LinkedListTimeRangeGroup.java]: I noticed that this class and ArrayListTimeRangeGroup share a fair amount of code. I would suggest that you use an abstract class to avoid duplicating the code.

ex:
abstract class AbstractListTimeRangeGroup implements TimeRangeGroup {
  protected List<TimeRange> allTimeRanges;

 ...
}

public class LinkedListTimeRangeGroup extends AbstractListTimeRangeGroup implements TimeRangeGroup {
  ...
}

File: portfolio/src/main/java/com/google/sps/data/LinkedListTimeRangeGroup.java [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+
+    while (iterator.hasNext()) {
+      TimeRange currentRange = iterator.next();
+
+      // If the current range is completely contained by the lastExaminedTimeRange,
+      // we need to remove the current range from the list as lastExaminedTimeRange
+      // will eventually replace it.
+      if (lastExaminedTimeRange.contains(currentRange)) {
+        iterator.remove();
+        continue;
```
Comment (jamesnam-google) [inline portfolio/src/main/java/com/google/sps/data/LinkedListTimeRangeGroup.java]: Consider the case where the initial internal time range list just has 1pm-2pm, and timeRange is 12pm-3pm. Then I think this line gets triggered, and the internal list stays at 1pm-2pm, rather than being updated to 12pm-3pm.

File: portfolio/src/main/java/com/google/sps/data/LinkedListTimeRangeGroup.java [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+      } else if (currentRange.end().isAfter(lastExaminedTimeRange.end())) {
+        // This is the case that lastExaminedTimeRange and the current time range do not overlap.
+        // If the current range from the original list ends after the last examined time range,
+        // add the time range pointed to by lastExaminedTimeRange to the new list before the current
+        // range,and change the lastExaminedTimeRange to point to the current range.
+        iterator.previous();
+        iterator.add(lastExaminedTimeRange);
+        iterator.next();
+        lastExaminedTimeRange = currentRange;
+      }
```
Comment (jamesnam-google) [inline portfolio/src/main/java/com/google/sps/data/LinkedListTimeRangeGroup.java:79]: Should we consider the else case? It can get triggered for example if the internal list was just 1pm-2pm, and timeRange is 3pm-4pm.

File: portfolio/src/main/java/com/google/sps/data/LinkedListTimeRangeGroup.java [hunk 6]
```diff
@@ -0,0 +1,187 @@
...
+        break;
+      }
+    }
+  }
+
+  /**
+   * Helper method for merging two time ranges. This method is package-private so that it can be
+   * tested.
+   */
+  static TimeRange mergeTwoTimeRanges(TimeRange a, TimeRange b) {
```
Comment (jamesnam-google) [inline portfolio/src/main/java/com/google/sps/data/LinkedListTimeRangeGroup.java]: This method can be a helper method defined in the abstract class.

File: portfolio/src/main/java/com/google/sps/data/LinkedListTimeRangeGroup.java [hunk 7]
```diff
@@ -0,0 +1,187 @@
...
+      newTimeRangeEnd = b.end();
+    } else {
+      newTimeRangeEnd = a.end();
+    }
+
+    return TimeRange.fromStartEnd(newTimeRangeStart, newTimeRangeEnd);
+  }
+
+  /** Returns an iterator for the list of all time ranges. */
+  public Iterator<TimeRange> getAllTimeRangesIterator() {
```
Comment (jamesnam-google) [inline portfolio/src/main/java/com/google/sps/data/LinkedListTimeRangeGroup.java]: The abstract class can have this method.

File: portfolio/src/main/java/com/google/sps/data/LinkedListTimeRangeGroup.java [hunk 8]
```diff
@@ -0,0 +1,187 @@
...
+   * <p>Another example for deleting overlapping time ranges: if [3 - 4] and [5 - 6] are in the
+   * original list, deleting [3:30 - 5:30] will result in two new ranges: [3 - 3:30] and [5:30 - 6].
+   */
+  public void deleteTimeRange(TimeRange timeRangeToDelete) {
+
+    ListIterator<TimeRange> iterator = allTimeRanges.listIterator();
+
+    while (iterator.hasNext()) {
+      TimeRange currentRange = iterator.next();
+      if (currentRange.overlaps(timeRangeToDelete)) {
```
Comment (jamesnam-google) [inline portfolio/src/main/java/com/google/sps/data/LinkedListTimeRangeGroup.java]: It might be simpler to continue in the opposite case, so you don't have to indent the inner block.

File: portfolio/src/main/java/com/google/sps/data/LinkedListTimeRangeGroup.java [hunk 9]
```diff
@@ -0,0 +1,187 @@
...
+   * overlap with any other existing time ranges.
+   *
+   * <p>For example, if the list contains [3:00 - 4:00] and [5:00 - 6:00], deleting [3:15 - 3:30]
+   * will result in [3 - 3:15] and [3:30 - 4] as new time ranges.
+   *
+   * <p>Another example for deleting overlapping time ranges: if [3 - 4] and [5 - 6] are in the
+   * original list, deleting [3:30 - 5:30] will result in two new ranges: [3 - 3:30] and [5:30 - 6].
+   */
+  public void deleteTimeRange(TimeRange timeRangeToDelete) {
+
```
Comment (jamesnam-google) [inline portfolio/src/main/java/com/google/sps/data/LinkedListTimeRangeGroup.java]: Unnecessary line break

File: portfolio/src/main/java/com/google/sps/data/LinkedListTimeRangeGroup.java [hunk 10]
```diff
@@ -0,0 +1,187 @@
...
+  /**
+   * Checks if a time range exists in the collection. For example, if [3:00 - 4:00] is in the
+   * collection, [3:00 - 3:30] is considered to exist as a time range in the collection. This method
+   * uses linear search to find the time ranges whose start time is before the target range's start
+   * and whose end time is after the target range's end. Then the method calls contains to see if
+   * the target range is contained within this current range.
+   */
+  public boolean hasTimeRange(TimeRange timeRangeToCheck) {
+
```
Comment (jamesnam-google) [inline portfolio/src/main/java/com/google/sps/data/LinkedListTimeRangeGroup.java]: Unnecessary line break

File: portfolio/src/test/java/com/google/sps/LinkedListTimeRangeGroupTest.java [hunk 11]
```diff
@@ -0,0 +1,557 @@
...
+import java.util.Iterator;
+import java.util.LinkedList;
+import java.util.List;
+import org.junit.Assert;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+
+@RunWith(JUnit4.class)
+public final class LinkedListTimeRangeGroupTest {
```
Comment (jamesnam-google) [inline portfolio/src/test/java/com/google/sps/LinkedListTimeRangeGroupTest.java]: Consider using a parameterized test, which runs the same test suite on LinkedListTimeRangeGroup as well as ArrayListTimeRangeGroup.

Reference:
https://junit.org/junit4/javadoc/4.12/org/junit/runners/Parameterized.html

Maybe we can name it as TimeRangeGroupTest

**Review by jamesnam-google — CHANGES_REQUESTED** [review]

File: portfolio/src/main/java/com/google/sps/data/LinkedListTimeRangeGroup.java [hunk 12]
```diff
@@ -0,0 +1,187 @@
...
+  // pair-wise disjoint at any moment.
+  public List<TimeRange> allTimeRanges;
+
+  /**
+   * Adds all the input time ranges to the list of all time ranges. Also sorts the list of all
+   * ranges in the constructor.
+   */
+  public LinkedListTimeRangeGroup(Collection<TimeRange> timeRanges) {
+    allTimeRanges = new LinkedList<TimeRange>();
+    timeRanges.forEach(
```
Comment (rgcase) [inline portfolio/src/main/java/com/google/sps/data/LinkedListTimeRangeGroup.java]: nit: You can just do `timeRanges.forEach(addTimeRange)`.

File: portfolio/src/main/java/com/google/sps/data/LinkedListTimeRangeGroup.java (hunk 7 above)
Comment (rgcase) [reply portfolio/src/main/java/com/google/sps/data/LinkedListTimeRangeGroup.java]: nit: It's named `.iterator()` in the other implementation and you should update yours to match, so it can be used in a foreach loop.

**Review by rgcase — CHANGES_REQUESTED** [review]

File: portfolio/src/main/java/com/google/sps/data/LinkedListTimeRangeGroup.java (hunk 2 above)
Comment (talvarez0507) [reply portfolio/src/main/java/com/google/sps/data/LinkedListTimeRangeGroup.java]: Done.

File: portfolio/src/main/java/com/google/sps/data/LinkedListTimeRangeGroup.java (hunk 3 above)
Comment (talvarez0507) [reply portfolio/src/main/java/com/google/sps/data/LinkedListTimeRangeGroup.java]: Done. I will change the ArrayList implementation to extend AbstractListTimeRangeGroup in another PR.

File: portfolio/src/main/java/com/google/sps/data/LinkedListTimeRangeGroup.java (hunk 4 above)
Comment (talvarez0507) [reply portfolio/src/main/java/com/google/sps/data/LinkedListTimeRangeGroup.java]: Fixed. I had to rewrite a bit of the code but all of these corner cases having to due with a single element or adding one element all the way at the end work fine now. Tests were also added for these cases.

File: portfolio/src/main/java/com/google/sps/data/LinkedListTimeRangeGroup.java (hunk 5 above)
Comment (talvarez0507) [reply portfolio/src/main/java/com/google/sps/data/LinkedListTimeRangeGroup.java:79]: The else case means that the currentRange and lastExaminedTimeRange do not overlap and that lastExaminedTimeRange is strictly later than currentRange. On some further iteration the else if will trigger and insert it into the list once it comes before something in the list so we don't have to worry about the else case.

File: portfolio/src/main/java/com/google/sps/data/LinkedListTimeRangeGroup.java (hunk 6 above)
Comment (talvarez0507) [reply portfolio/src/main/java/com/google/sps/data/LinkedListTimeRangeGroup.java]: Done.

File: portfolio/src/main/java/com/google/sps/data/LinkedListTimeRangeGroup.java (hunk 7 above)
Comment (talvarez0507) [reply portfolio/src/main/java/com/google/sps/data/LinkedListTimeRangeGroup.java]: Done.

File: portfolio/src/main/java/com/google/sps/data/LinkedListTimeRangeGroup.java (hunk 8 above)
Comment (talvarez0507) [reply portfolio/src/main/java/com/google/sps/data/LinkedListTimeRangeGroup.java]: Done.

File: portfolio/src/main/java/com/google/sps/data/LinkedListTimeRangeGroup.java (hunk 9 above)
Comment (talvarez0507) [reply portfolio/src/main/java/com/google/sps/data/LinkedListTimeRangeGroup.java]: Removed.

File: portfolio/src/main/java/com/google/sps/data/LinkedListTimeRangeGroup.java (hunk 10 above)
Comment (talvarez0507) [reply portfolio/src/main/java/com/google/sps/data/LinkedListTimeRangeGroup.java]: Removed.

File: portfolio/src/test/java/com/google/sps/LinkedListTimeRangeGroupTest.java (hunk 11 above)
Comment (talvarez0507) [reply portfolio/src/test/java/com/google/sps/LinkedListTimeRangeGroupTest.java]: Done.

File: portfolio/src/main/java/com/google/sps/data/LinkedListTimeRangeGroup.java (hunk 12 above)
Comment (talvarez0507) [reply portfolio/src/main/java/com/google/sps/data/LinkedListTimeRangeGroup.java]: I was not able to make this work. I got the error message "cannot find symbol addTimeRange" even though it is a method in the class.

File: portfolio/src/main/java/com/google/sps/data/AbstractListTimeRangeGroup.java [hunk 13]
```diff
@@ -0,0 +1,43 @@
...
+import java.util.List;
+
+abstract class AbstractListTimeRangeGroup implements TimeRangeGroup {
+  protected List<TimeRange> allTimeRanges;
+
+  /**
+   * Helper method for merging two time ranges. This method is package-private so that it can be
+   * tested.
+   */
+  public static TimeRange mergeTwoTimeRanges(TimeRange a, TimeRange b) {
```
Comment (jamesnam-google) [inline portfolio/src/main/java/com/google/sps/data/AbstractListTimeRangeGroup.java]: I think this method can be protected instead of public.

File: portfolio/src/main/java/com/google/sps/data/AbstractListTimeRangeGroup.java [hunk 14]
```diff
@@ -0,0 +1,43 @@
...
+      newTimeRangeEnd = b.end();
+    } else {
+      newTimeRangeEnd = a.end();
+    }
+
+    return TimeRange.fromStartEnd(newTimeRangeStart, newTimeRangeEnd);
+  }
+
+  /** Returns an iterator for the list of all time ranges. */
+  public Iterator<TimeRange> iterator() {
```
Comment (jamesnam-google) [inline portfolio/src/main/java/com/google/sps/data/AbstractListTimeRangeGroup.java:41]: (Optional) You can add the @Override annotation here (capital O, which somehow is not capitalized when I submit the comment on GitHub).

Reference:
https://docs.oracle.com/javase/tutorial/java/annotations/predefined.html

File: portfolio/src/main/java/com/google/sps/data/ArrayListTimeRangeGroup.java [hunk 15]
```diff
@@ -0,0 +1,146 @@
...
+        (range) -> {
+          addTimeRange(range);
+        });
+  }
+
+  /**
+   * Adds a new time range to the list. If the time range to add overlaps with any existing time
+   * range, the overlapping time ranges will be merged.
+   */
+  public void addTimeRange(TimeRange timeRange) {
```
Comment (jamesnam-google) [inline portfolio/src/main/java/com/google/sps/data/ArrayListTimeRangeGroup.java:24]: You can consider adding the override annotation

File: portfolio/src/main/java/com/google/sps/data/ArrayListTimeRangeGroup.java [hunk 16]
```diff
@@ -0,0 +1,146 @@
...
+/** Models an implementation of the TimeRangeGroup model using ArrayList. */
+public class ArrayListTimeRangeGroup extends AbstractListTimeRangeGroup implements TimeRangeGroup {
+
+  /**
+   * Adds all the input time ranges to the list of all time ranges. Also sorts the list of all
+   * ranges in the constructor.
+   */
+  public ArrayListTimeRangeGroup(Collection<TimeRange> timeRanges) {
+    allTimeRanges = new ArrayList<TimeRange>();
+    timeRanges.forEach(
```
Comment (jamesnam-google) [inline portfolio/src/main/java/com/google/sps/data/ArrayListTimeRangeGroup.java]: Consider pulling out this forEach into a helper method. I think it can be a protected method declared in the abstract class. Then here you can invoke it.

(As a result, you may have to declare addTimeRange in the abstract method.)

File: portfolio/src/test/java/com/google/sps/TimeRangeGroupTest.java [hunk 17]
```diff
@@ -0,0 +1,669 @@
...
+import org.junit.runners.Parameterized.Parameters;
+
+@RunWith(value = Parameterized.class)
+public class TimeRangeGroupTest {
+  @Parameters
+  public static List parameters() {
+    return Arrays.asList(ArrayListTimeRangeGroup.class, LinkedListTimeRangeGroup.class);
+  }
+
+  private Class timeRangeGroupClass;
```
Comment (jamesnam-google) [inline portfolio/src/test/java/com/google/sps/TimeRangeGroupTest.java]: I think you can do,

@Parameter Class timeRangeGroupClass;

Then you won't need the TimeRangeGroupTest constructor below.

File: portfolio/src/main/java/com/google/sps/data/LinkedListTimeRangeGroup.java [hunk 18]
```diff
@@ -0,0 +1,157 @@
...
+ */
+public class LinkedListTimeRangeGroup extends AbstractListTimeRangeGroup implements TimeRangeGroup {
+
+  /**
+   * Adds all the input time ranges to the list of all time ranges. Also sorts the list of all
+   * ranges in the constructor.
+   */
+  public LinkedListTimeRangeGroup(Collection<TimeRange> timeRanges) {
+    allTimeRanges = new LinkedList<TimeRange>();
+    timeRanges.forEach(
```
Comment (jamesnam-google) [inline portfolio/src/main/java/com/google/sps/data/LinkedListTimeRangeGroup.java]: As I suggested earlier, I think you can just invoke the helper method here for forEach.

File: portfolio/src/main/java/com/google/sps/data/LinkedListTimeRangeGroup.java [hunk 19]
```diff
@@ -0,0 +1,157 @@
...
+    // linked list can be confusing. Each case should be handled differently
+    // so using iterator.hasNext() and currentRange=iterator.next() in each
+    // iteration of this loop would make things harder to keep track of.
+    while (true) {
+
+      // If the current range is completely contained by the lastExaminedTimeRange,
+      // we need to remove the current range from the list as lastExaminedTimeRange
+      // will eventually replace it.
+      if (lastExaminedTimeRange.contains(currentRange)
+          && !lastExaminedTimeRange.equals(currentRange)) {
```
Comment (jamesnam-google) [inline portfolio/src/main/java/com/google/sps/data/LinkedListTimeRangeGroup.java:57]: Do we need the ! equals check?

File: portfolio/src/main/java/com/google/sps/data/LinkedListTimeRangeGroup.java [hunk 20]
```diff
@@ -0,0 +1,157 @@
...
+        iterator.remove();
+        if (iterator.hasNext()) {
+          currentRange = iterator.next();
+        }
+      }
+
+      // The case for when two time ranges need to be merged.
+      // Similar to the case above, it will eventually be replaced so we can
+      // remove currentRange
+      if (lastExaminedTimeRange.overlaps(currentRange)
```
Comment (jamesnam-google) [inline portfolio/src/main/java/com/google/sps/data/LinkedListTimeRangeGroup.java:63]: Can this be an else if? If so, then we won't need the ! contains check.

**Review by jamesnam-google — APPROVED** [review]
Approving the change with comments

File: portfolio/src/test/java/com/google/sps/TimeRangeGroupTest.java [hunk 21]
```diff
@@ -0,0 +1,669 @@
...
+
+  /** Tests for adding a time range that entirely contains an existing one. */
+  @Test
+  public void testAddSupersetRange() {
+    // Time Ranges:      |-A-|
+    // To add:       |-----B-----|
+    // Result:       |-----------|
+    Instant timeRangeOneStart = Instant.now();
+    Instant timeRangeOneEnd = timeRangeOneStart.plusSeconds(3000);
+    Instant timeRangeTwoStart = timeRangeOneStart.plusSeconds(500);
```
Comment (hollyyuqizheng) [inline portfolio/src/test/java/com/google/sps/TimeRangeGroupTest.java]: timeRangeTwoStart should probably be `timeRangeOneStart.minusSeconds(500)` ?

File: portfolio/src/test/java/com/google/sps/TimeRangeGroupTest.java [hunk 22]
```diff
@@ -0,0 +1,669 @@
...
+
+    TimeRange timeRangeOne = TimeRange.fromStartEnd(timeRangeOneStart, timeRangeOneEnd);
+
+    List<TimeRange> originalTimeRanges = Arrays.asList(timeRangeOne);
+    for (TimeRange timeRange : originalTimeRanges) {
+      timeRangeGroup.addTimeRange(timeRange);
+    }
+    List<TimeRange> expected = Arrays.asList(timeRangeOne);
+
+    timeRangeGroup.addTimeRange(timeRangeOne);
```
Comment (hollyyuqizheng) [inline portfolio/src/test/java/com/google/sps/TimeRangeGroupTest.java:290]: Hmmm what if you try to construct a new TimeRange instance called `timeRangeTwo`, and initialize this new time range with the exact start and end times of timeRangeOne? I think in this case the test passes for array list time range group, at least it did when I tried it on my own.

File: portfolio/src/main/java/com/google/sps/data/AbstractListTimeRangeGroup.java (hunk 13 above)
Comment (talvarez0507) [reply portfolio/src/main/java/com/google/sps/data/AbstractListTimeRangeGroup.java]: Done.

File: portfolio/src/main/java/com/google/sps/data/AbstractListTimeRangeGroup.java (hunk 14 above)
Comment (talvarez0507) [reply portfolio/src/main/java/com/google/sps/data/AbstractListTimeRangeGroup.java:41]: Done.

File: portfolio/src/main/java/com/google/sps/data/ArrayListTimeRangeGroup.java (hunk 15 above)
Comment (talvarez0507) [reply portfolio/src/main/java/com/google/sps/data/ArrayListTimeRangeGroup.java:24]: Done.

File: portfolio/src/main/java/com/google/sps/data/ArrayListTimeRangeGroup.java (hunk 16 above)
Comment (talvarez0507) [reply portfolio/src/main/java/com/google/sps/data/ArrayListTimeRangeGroup.java]: Done.

File: portfolio/src/test/java/com/google/sps/TimeRangeGroupTest.java (hunk 17 above)
Comment (talvarez0507) [reply portfolio/src/test/java/com/google/sps/TimeRangeGroupTest.java]: Done.

File: portfolio/src/main/java/com/google/sps/data/LinkedListTimeRangeGroup.java (hunk 18 above)
Comment (talvarez0507) [reply portfolio/src/main/java/com/google/sps/data/LinkedListTimeRangeGroup.java]: Done.

File: portfolio/src/main/java/com/google/sps/data/LinkedListTimeRangeGroup.java (hunk 19 above)
Comment (talvarez0507) [reply portfolio/src/main/java/com/google/sps/data/LinkedListTimeRangeGroup.java:57]: If we remove the !equals then when we try to add an event that already exists it is removed and never added.

File: portfolio/src/main/java/com/google/sps/data/LinkedListTimeRangeGroup.java (hunk 20 above)
Comment (talvarez0507) [reply portfolio/src/main/java/com/google/sps/data/LinkedListTimeRangeGroup.java:63]: I think this can't be an else if because the if statement that comes before this and the else if statement after this are not mutually exclusive.

File: portfolio/src/test/java/com/google/sps/TimeRangeGroupTest.java (hunk 21 above)
Comment (talvarez0507) [reply portfolio/src/test/java/com/google/sps/TimeRangeGroupTest.java]: Fixed.

File: portfolio/src/test/java/com/google/sps/TimeRangeGroupTest.java (hunk 22 above)
Comment (talvarez0507) [reply portfolio/src/test/java/com/google/sps/TimeRangeGroupTest.java:290]: Seems like I just had outdated code, ArrayListTimeRangeGroup passes this test just fine now.

**Review by rgcase — APPROVED** [review]
